lib/streamlit/runtime/state/query_params.py

```diff
@@ -0,0 +1,126 @@
+
+    def _getitem(self, key: str) -> str:
+        try:
+            value = self._query_params[key]
+            if isinstance(value, list):
+                if len(value) == 0:
+                    return ""
+                else:
+                    return value[-1]
+            return value
+        except:
+            raise KeyError(_missing_key_error_message_query_params(key))
+
+    def _setitem(self, key: str, value: Union[str, List[str]]) -> None:
+        if isinstance(value, list):
+            self._query_params[key] = [str(item) for item in value]
+        else:
+            self._query_params[key] = str(value)
+        self._send_query_param_msg()
+
+    def __setitem__(self, key: str, value: str) -> None:
+        self._setitem(key, value)
+
+    def get_all(self, key: str) -> List[str]:
+        try:
+            if key not in self._query_params:
+                return []
+            query_params = self._query_params[key]
+            return query_params if isinstance(query_params, list) else [query_params]
+        except KeyError:
+            raise KeyError(_missing_key_error_message_query_params(key))
+
+    def __getattr__(self, key: str) -> str:
+        try:
+            return self._getitem(key)
+        except KeyError:
+            raise AttributeError(_missing_key_error_message_query_params(key))
+
+    def __setattr__(self, key: str, value: str) -> None:
+        self._setitem(key, value)
+
+    def __delattr__(self, key: str) -> None:
+        try:
+            self._delitem(key)
+        except KeyError:
+            raise AttributeError(_missing_key_error_message_query_params(key))
+
+    def _delitem(self, key):
+        if key in self._query_params:
+            del self._query_params[key]
+            self._send_query_param_msg()
+        else:
+            raise KeyError(_missing_key_error_message_query_params(key))
+
+    def __contains__(self, key: str) -> bool:  # type: ignore[override]
+        return key in self._query_params
+
+    def __len__(self) -> int:
+        return len(self._query_params)
+
+    def _send_query_param_msg(self) -> None:
+        # Avoid circular imports
+        from streamlit.commands.experimental_query_params import _ensure_no_embed_params
+        from streamlit.runtime.scriptrunner import get_script_run_ctx
+
+        ctx = get_script_run_ctx()
+        if ctx is None:
+            return
+        msg = ForwardMsg()
+        msg.page_info_changed.query_string = _ensure_no_embed_params(
+            self._query_params, ctx.query_string
+        )
+        ctx.query_string = msg.page_info_changed.query_string
+        ctx.enqueue(msg)
+
+    def clear(self) -> None:
+        self._query_params.clear()
+        self._send_query_param_msg()
+
+    def __delitem__(self, key: str) -> None:
+        self._delitem(key)
+
+    def to_dict(self) -> Dict[str, Union[List[str], str]]:
+        return self._query_params
+
+
+def _missing_key_error_message_query_params(key: str) -> str:
```

Comment:
nit: I feel like this name is too verbose. There's no need to have the `_query_params` suffix in it because the function is local to this file so it's obvious what the error message is for